executor/infoschema_reader.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/privilege"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"sort"
+)
+
+// InfoschemaReaderExec executes infoschema information retrieving
+type InfoschemaReaderExec struct {
+	baseExecutor
+	iter      kv.Iterator
+	table     *model.TableInfo
+	columns   []*model.ColumnInfo
+	ChunkList *chunk.List
+	ChunkIdx  int
+}
+
+// Open implements the Executor Open interface.
+func (e *InfoschemaReaderExec) Open(ctx context.Context) error {
+	e.iter = nil
+	e.ChunkList = nil
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *InfoschemaReaderExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.GrowAndReset(e.maxChunkSize)
+	if e.ChunkList == nil {
+		e.ChunkList = chunk.NewList(retTypes(e), e.initCap, e.maxChunkSize)
+		mutableRow := chunk.MutRowFromTypes(retTypes(e))
+		rows, err := e.getRows(e.ctx, e.columns)
```

Comment:
How about change to `err := e.getRows(e.ctx, e.columns, chk)`, and then remove `ChunkList ` and `ChunkIdx `.

Reply:
Due to the other infoschema table, if put the chk in `getRow()`, it would need a logic to append data to chk depending on the `field type` which get by `DataForXXX` func, which seems not clean; if i put the 'append data to chk logic' into `DataForXXX` func, then i would change every  `DataForXXX` func, that is not good. With ChunkList and ChunkIdx, i can reuse the  'append data to chk logic' by `chunkList.AppendRow()`, which would not cause much change.

Reply:
Why not cache `rows` and `rowIdx` in `InfoschemaReaderExec` dirrectly?

Mostly, the parameter of `chk` is reused for cache and avoid re-allocate memory.

Suppose the length of `rows` is 10240, currently, this code will need to allocate 10240 size for `chunkList`,  and without `chunkList`, we can only allocate 1024 size for the `chk` and reuse it for cache ( Suppose e.MaxChunkSize = 1024).

Reply:
Great! got it.